R/compile_sim_cpp.R

```diff
@@ -209,7 +209,18 @@ compile_sim_cpp <- function(
       cpp_code[idx13] <- paste0('  comb["obs"] = obs;\n');
       cpp_code[idx14] <- "  std::vector<double> obs;"
     } else {
-      if(is.null(obs$cmt)) obs$cmt <- 1
+      if(is.null(obs$cmt) && is.null(obs$variable)) {
+        stop("Either a `cmt` or `variable` element has to be specified in the `obs` argument.")
+      }
+      if(!is.null(obs$cmt)) { ## `cmt` is specified
+        if(length(obs$scale) != length(obs$cmt)) {
+          stop("Provided `scale` vector is not the same length as observation compartments (`cmt`) vector. Please update `obs` argument.")
+        }
+      } else { ## `variable` is used
+        if(length(obs$scale) != length(obs$variable)) {
+          stop("Provided `scale` vector is not the same length as observation `variable` vector. Please update `obs` argument.")
+        }
+      }
       for(k in 1:length(obs$cmt)) {
```

Comment:
If I read this right, we'd require either `cmt` or `variable`. But if `variable` is provided and `cmt` is null, then won't the for loop below (`for k in 1:length(obs$cmt)`) fail?

Reply:
Right. I think that whole for-loop could be moved into the if-statement on line 215. I.e. I don't think it's necessary to include this loop when the observation is specified using a variable instead of cmt. The only important grep label for `variable` is idx2, and that is handled on line 233-235. But not entirely sure, would need to test.

Another option is to move that for loop out of this block and one level up. I.e. add it as `else` to the if-statement on line 233-235?
```
if(!is.null(obs$variable) && length(obs$variable) > 0) {
  cpp_code[idx12] <- parse_obs_types(obs)
} else {
   for(k in 1:length(obs$cmt)) {
        cpp_code[idx5] <- paste0(cpp_code[idx5], "\n    scale", k," = ", obs$scale[k], ";")
        cpp_code[idx7] <- paste0(cpp_code[idx7], "\n      scale", k," = ", obs$scale[k], ";")
        cpp_code[idx8] <- cov_scale
        cpp_code[idx12] <- paste0(cpp_code[idx12], "\n      obs",k,".insert(obs",k,".end(), tmp.y[k][", obs$cmt[k]-1,"] / scale", k,");")
        cpp_code[idx13] <- paste0(cpp_code[idx13], '\n  comb["obs', k,'"] = obs', k,';');
        cpp_code[idx14] <- paste0(cpp_code[idx14], "\n  std::vector<double> obs",k,";")
  }
}
```

Reply:
Coming back to this, I think in most cases this won't behave as expected, because `new_ode_model()` alters the `obs` object before passing it to `compile_sim_cpp()`. See the below example where we have a scale/variable length mismatch and no `cmt`. We should expect an error, but we don't get one:

``` r
## define parameters
vars <- c("CONC", "METAB", "METAB2")
pk1 <- new_ode_model(
  code = "dAdt[1] = -(CL/V)*A[1]; CONC = 1000*A[1]/V; METAB = CONC/2; METAB2 = CONC * t;",
  obs = list(variable = vars, scale = c(1, 1)), # mismatch vars/scale length; no cmt
  declare_variables = vars,
  cpp_show_code = F
)
#> Warning in parse_obs_types(obs): Provided `scale` vector not same length as
#> `variable` vector.
```

That's because `new_ode_model()` adds a `cmt` element to the `obs` object, so we never make it to these error conditions.

https://github.com/InsightRX/PKPDsim/blob/08677e82f59e37882a2bd76a04d39b4d28fe849b/R/new_ode_model.R#L193

I think it might be a good idea to come up with a more comprehensive set of test cases for this behavior before making more changes to the code. Maybe a brief meeting would help?

Reply:
Coming back to this, I think in most cases this won't behave as expected, because `new_ode_model()` alters the `obs` object before passing it to `compile_sim_cpp()`. See the below example where we have a scale/variable length mismatch and no `cmt`. We should expect an error, but we don't get one:

``` r
## define parameters
vars <- c("CONC", "METAB", "METAB2")
pk1 <- new_ode_model(
  code = "dAdt[1] = -(CL/V)*A[1]; CONC = 1000*A[1]/V; METAB = CONC/2; METAB2 = CONC * t;",
  obs = list(variable = vars, scale = c(1, 1)), # mismatch vars/scale length; no cmt
  declare_variables = vars,
  cpp_show_code = F
)
#> Warning in parse_obs_types(obs): Provided `scale` vector not same length as
#> `variable` vector.
```

That's because `new_ode_model()` adds a `cmt` element to the `obs` object, so we never make it to these error conditions.

https://github.com/InsightRX/PKPDsim/blob/08677e82f59e37882a2bd76a04d39b4d28fe849b/R/new_ode_model.R#L193

I think it might be a good idea to come up with a more comprehensive set of test cases for this behavior before making more changes to the code. Maybe a brief meeting would help?